File: ios/Runner/MediaPlugin/MediaPlugin.m
//
//  MediaPlugin.m
//  Runner
//
//  Created by chihara on 2018/08/30.
//  Copyright © 2018年 The Chromium Authors. All rights reserved.
//

#import "MediaPlugin.h"
#import <MediaPlayer/MediaPlayer.h>

@interface FLTMediaPlugin ()<FlutterStreamHandler>
@end

@implementation FLTMediaPlugin {
    FlutterEventSink _eventSink;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    FLTMediaPlugin* instance = [[FLTMediaPlugin alloc] init];
    
    FlutterMethodChannel* channel = [FlutterMethodChannel methodChannelWithName:@"com.example.player/media" binaryMessenger:[registrar messenger]];
    
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSLog(@"handleMethodCall:%@", call.method);
    if ([@"getSongs" isEqualToString:call.method]) {
        NSMutableArray<NSMutableDictionary*>* list = [NSMutableArray<NSMutableDictionary*> array];
        MPMediaQuery* query = [MPMediaQuery songsQuery];
        NSArray* items = query.items;
        for (MPMediaItem* item in items) {
            NSMutableDictionary* dic = [NSMutableDictionary<NSString*, NSObject*> dictionary];
            dic[@"title"] = item.title;
            dic[@"artist"] = item.artist;
            dic[@"album"] = item.albumTitle;
            dic[@"albumArt"] = item.artwork;
            dic[@"track"] = [NSNumber numberWithInteger:item.albumTrackNumber];
            dic[@"duration"] = [NSNumber numberWithInteger:item.playbackDuration];
            [list addObject:dic];
        }
        result(list);
    }
    if ([@"getAlbums" isEqualToString:call.method]) {
        NSMutableArray<NSMutableDictionary*>* list = [NSMutableArray<NSMutableDictionary*> array];
        MPMediaQuery* query = [MPMediaQuery albumsQuery];
        NSArray* items = query.items;
        for (MPMediaItem* item in items) {
            NSMutableDictionary* dic = [NSMutableDictionary<NSString*, NSObject*> dictionary];
            dic[@"title"] = item.title;
            dic[@"artist"] = item.artist;
            dic[@"albumArt"] = item.artwork;
            dic[@"numOfTracks"] = [NSNumber numberWithInteger:item.albumTrackCount];
            [list addObject:dic];
        }
        result(list);
    }
}

@end

File: ios/Runner/MediaPlugin/MediaPlugin.swift
//
//  MediaPlugin.swift
//  Runner
//
//  Created by 千原　寛之 on 2018/09/03.
//  Copyright © 2018年 The Chromium Authors. All rights reserved.
//

import Foundation
import MediaPlayer
import Flutter

public class MediaPlugin: NSObject, FlutterPlugin {
    
    private let player = MPMusicPlayerController.applicationMusicPlayer
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        print("register")
        let channel = FlutterMethodChannel(
            name: "com.example.player/media",
            binaryMessenger: registrar.messenger()
        )
        let instance = MediaPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        debugPrint("handle:%s", call.method)
        switch call.method {
        case "getSongs":
            let list = NSMutableArray.init()
//            let items = MPMediaQuery.songs().items
//            for item in items! {
//                let dic = NSMutableDictionary()
//                dic["id"] = item.persistentID
//                dic["title"] = item.title
//                dic["artist"] = item.artist
//                dic["album"] = item.albumTitle
//                dic["albumArt"] = item.artwork
//                dic["track"] = item.albumTrackNumber
//                dic["duration"] = item.playbackDuration
//                list.add(dic)
//            }
            createDummySongs(list: list)
            result(list)
        case "getAlbums":
            let list = NSMutableArray.init()
//            let items = MPMediaQuery.albums().items
//            for item in items! {
//                let dic = NSMutableDictionary()
//                dic["id"] = item.albumPersistentID
//                dic["title"] = item.title
//                dic["artist"] = item.artist
//                dic["albumArt"] = item.artwork
//                dic["numOfTracks"] = item.albumTrackCount
//                list.add(dic)
//            }
            createDummyAlbums(list: list)
            result(list)
        case "play":
            if (MPMusicPlaybackState.playing == player.playbackState) {
                player.pause()
            } else {
                let song = Song.fromDic(dic: call.arguments as! NSDictionary)
                let query = MPMediaQuery.songs()
                query.addFilterPredicate(MPMediaPropertyPredicate(value: song.id, forProperty: MPMediaItemPropertyPersistentID))
                player.setQueue(with: query)
                player.play()
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func createDummySongs(list: NSMutableArray) {
        for num in 1...25 {
            let song = Song(
                id: 0,
                title: "item.title\(num)",
                artist: "item.artist\(num)",
                album: "item.albumTitle\(num)",
                albumArt: "",
                track: num,
                duration: UInt64(100000 * num))
            list.add(song.toDic())
        }
    }
    
    private func createDummyAlbums(list: NSMutableArray) {
        for num in 1...11 {
            let album = Album(
                id: 0,
                title: "item.title\(num)",
                artist: "item.artist\(num)",
                albumArt: "",
                numOfTracks: num)
            list.add(album.toDic())
        }
    }
}
